features/storage/kvstore/include/KVStore.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2020 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------- Includes -----------------------------------------------------------
+
+#include "KVStore.h"
+namespace mbed {
+
+
+bool KVStore::is_valid_key(const char *key) const
+{
+    if (!key || !strlen(key) || (strlen(key) > MAX_KEY_SIZE)) {
+        return false;
+    }
+
+    if (strpbrk(key, " */?:;\"|<>\\")) {
+        return false;
+    }
+    return true;
+}
+
+bool KVStore::_is_valid_flags(InitModeFlags flags)
+{
+    // Check that only valid bits are here at all
+    if ((~(InitModeFlags::AllFlags) & flags) == InitModeFlags::NoFlags) {
+        return false;
+    }
+
+    // Need at least one of the Read or Write bits set
+    if ((flags & InitModeFlags::ReadWrite) == InitModeFlags::NoFlags) {
+        return false;
+    }
+
+    // Check for only one set of the WriteOpenFlags
+    const uint32_t wo_flags = static_cast<uint32_t>(flags & InitModeFlags::WriteOpenFlags);
+    // Fails for zero, the conditional afterwards finds this
+    if (wo_flags & (wo_flags - 1)) {
+        return false;
+    }
+    // Only allow blank Append, etc. if no writing
+    if (((flags & InitModeFlags::Write) == InitModeFlags::Write) && (wo_flags == 0)) {
```

Comment:
This sort of verbosity grates, particularly the repetition. I feel it should be fine to say `if (flags & X)`. But it isn't :(

Digging into it, I see there is some confusion about "bitmask types". See issue 3092 at https://cplusplus.github.io/LWG/lwg-unresolved.html 

And previously there have been notes about stuff assuming what I was thinking, eg: https://cplusplus.github.io/LWG/issue2659

In the latter they settled on `if ((flags & options::X) != options::none)`, which is mildly better IMO, if only because it's not repetitive. As the 3092 issue points out though, that could be shortened to `if ((flags & options::X) != options())`, which I quite like, visualising the empty bitset.

Still, would have been nice to get the contextual conversion to bool - I see one maniac on the web trying to go fully type safe, distinguishing the "bitmask" type from the "bit value" type, so you get contextual conversion to bool on the result of `&` and aren't allowed to accidentally do `==` instead of `&`: https://dalzhim.github.io/2017/08/11/Improving-the-enum-class-bitmask/

I find that actually quite tempting. BSD license...